### ISSUE TITLE
RFC: Consider changing how we accept third-party plugins

### ISSUE THREAD

**Issue by laurent22** [laurent22/joplin#9582]
Currently, the Joplin application gets the list of plugins from our [GitHub repository](https://github.com/joplin/plugins), and that repository is automatically built based on the third-party plugins that have been published to npm. We generally don't do any checks on these plugins except for those that are marked as "recommended", which are those developed by the team or long term contributors.

In order to make it safer for users to install plugins, it would be good to have a review process. Many plugins are relatively simple and we have perhaps 3 or 4 updates per week so it should be doable. The advantage is that we can guarantee that all reviewed plugins are safe to install, which would be interesting to users who no longer need to limit themselves to recommended plugins only.

To minimise changes to the existing process, we could integrate the review process in that way:

- Publishing to npm will no longer be relevant
- Instead, we would check the code directly on the repository
    - The code itself
    - But also the dependencies - we should only accept dependencies to well-known, safe packages
- If all is good, we record the repository url and commit that we've reviewed
- When building the plugin repository, we would automatically:
    - Fetch the third-party repositories at the previously recorded commit
    - Build the plugin
    - Add the .jpl to our plugin repository

Once we have the .jpl the rest of our process will be identical.

## Questions

- What to do about plugins that we have not yet reviewed - should we include them and mark them as "not reviewed". Or exclude them completely?
- How about plugins that are not published to any public repository? (not sure if they exist)
- For review, we should have some tool to generate a diff between the last reviewed commit and the latest one
- We should probably have a process to allow the plugin developer to tell us when their plugin is ready for review, otherwise we may fetch a commit that's still under development

**Comment by Shbhom** [comment]
Hi, I'm Shubhom Srivastava, I liked the problem this project is trying to solve. But I'm thinking how are going to do the code review. 
Since, I'm still in college I don't know different code review tools used, in my current internship we usually sit in a meet with a senior and do the code review.
what I'm thinking to do for this problem is building a CLI tool like Heroku which has similar commands to git, using this we can do get two things done
1. VCS support for each change in code for the 3rd party plugin
2. we can create a clone of serverless deployment sites like vercel which builds the code they get in an isolated docker container. so our requirement for a trusted server to build plugins will also be full filed.
3. and we can use middlewares between the plugin authors pc and our git server which would apply logic for code review.

Currently working on the code review part, @laurent22 can you review this proposal ?

**Comment by rtfmkiesel** [comment]
To give my 2 cents here as the one who found the XSS in the Outline plugin last year...

I had to [wait](https://github.com/cqroot/joplin-outline/issues/100) for the developer to add a [SECURITY.md](https://docs.github.com/en/code-security/getting-started/adding-a-security-policy-to-your-repository) to the plugin repo before I was able to contact him. You could add this as a requirement (to get listed) so that in case of a vulnerability, the researcher knows how to contact the developer(s). They can then either use GitHubs internal security tracker or just specify an email.

The other option is a central way of reporting issues, maybe through a SECURITY.md in the [joplin/plugins](https://github.com/joplin/plugins) repo. The benefit there is that you would automatically know of vulnerabilities, and in severe cases could maybe stop the distribution of the plugin through the store temporarily until the issue is resolved. But this of course would generate work for you. 

Happy to give more input into this if needed.

**Comment by laurent22** [comment]
@rtfmkiesel, yes I think that makes sense. And in a way the security of plugins fall on us anyway because ideally (and that's not currently the case) it should be possible to run any plugin safely. 

In fact the security report link in mobile already link to our own security.md file. And indeed if a plugin has a severe vulnerability and the developer is no longer around we'd probably have to take the plugin out of the repository.

We don't have a clear process in place for this at the moment and I guess it could be addressed as part of this RFC.

**Comment by Ehtesham-Zahid** [comment]
Thanks for starting this discussion. I’ve recently been looking into the plugin loading and execution flow, so I wanted to share a few thoughts on the questions raised here.

**Plugins that haven't been reviewed**

It might be better to include them but clearly mark them as **“not reviewed”** instead of excluding them completely. This keeps the ecosystem open while still giving users some context about the level of trust. For example:

* **Reviewed** – passed manual/automated review
* **Not reviewed** – available but install with caution
* **Recommended** – maintained by the core team or long-term contributors

**Plugins without a public repository**

Requiring a **public repository** would probably make sense. Without it, reviewing the code, checking dependencies, or comparing updates becomes difficult and removes transparency for both maintainers and users.

**Diffing changes between reviews**

Instead of manually comparing versions, the review system could store the **last reviewed commit hash** for each plugin. When a developer requests a new review, tooling could generate a diff between the stored commit and the new one.

The generated report could highlight changes that are more relevant from a security perspective, for example:

* modifications in source files
* changes to `package.json` (especially new dependencies)
* changes to build scripts or install hooks
* usage of sensitive APIs (`eval`, dynamic imports, network access, filesystem access)

Rather than showing a raw Git diff, the tool could summarize these changes so reviewers can quickly focus on potentially risky updates.

**Letting developers request reviews**

One option could be letting plugin developers open a **review request issue (or PR)** with the repository URL and commit they want reviewed. The system could then record that commit as the trusted one used during the plugin build process.

**One additional thought**

Since users ultimately install compiled `.jpl` files, it might help if these were always **built by the official pipeline from the reviewed commit**, rather than uploaded by plugin authors. This would ensure that the distributed package matches the reviewed source.

Overall, reviewing repositories and building plugins centrally seems like a good balance between improving security while keeping the plugin ecosystem open.

For context, I’ve been exploring the Joplin plugin architecture while preparing a **GSoC 2026 proposal focused on plugin security**, so this topic is particularly interesting to me.

**Comment by Ehtesham-Zahid** [comment]
TL;DR: Include plugins marked as “not reviewed,” require public repos, generate diffs from last reviewed commit highlighting security-relevant changes, let developers request reviews via issues/PRs, and always build .jpl files from reviewed commits. Centralized review + build keeps the ecosystem open while improving security. Open for discussion.

**Comment by ahmed-n-abdeltwab** [comment]
I have an idea: I could design an ML model that searches for suspect code, such as binary or string injections, or the execution of Node packages responsible for OS or network operations. We could continue adding parameters to enable the model to detect whether a library is safe or malicious. I am raising this point because well-known dependencies are not necessarily secure; they could contain vulnerabilities, similar to those previously found in React or Angular. Just because a software is popular does not mean it is good. We should implement an ML model that learns from multiple cases and parameters to determine if a software is trustworthy or harmful. additionally we need to implement a proxy layer that's stand before the dependencies and Joplin application so we could feel 100% sure that's if the code has something it's wouldn't affect the applications

**Comment by roy-stack-web** [comment]
Hi,
I’m interested in working on this for GSoC 2026.

While going through the RFC, I was thinking that along with reviewing plugins at a specific commit, it might help to have a simple automated check that highlights important changes between the last reviewed commit and the new one : like new dependencies, package.json updates, or usage of sensitive APIs.

This could make it easier for reviewers to quickly spot risky changes.

**Comment by laurent22** [comment]
@roy-stack-web, indeed some high level summary would make sense to know what we're dealing with and ensure we're not missing anything. One example would be dependencies - we could accidentally overlook a change in package.json, but with an overview that would be less likely

**Comment by roy-stack-web** [comment]
@laurent22, That makes sense. To make sure changes like dependencies aren’t missed, the summary could explicitly separate things like package.json updates and newly added dependencies from the rest of the code changes.

For example : 
Showing sections like “New dependencies”, “Modified dependencies”, and “Changes in source files” could help reviewers quickly focus on higher-risk updates first.

**Comment by roy-stack-web** [comment]
While exploring the plugin loading flow, I noticed that in `loadPluginFromPath`, plugins are read and passed directly to `loadPlugin` without any intermediate checks.

It might be useful to introduce a lightweight verification step here that scans the plugin’s code and dependencies and generates a small summary before loading. This could help catch potential risks early and also integrate well with the review pipeline discussed above.

**Comment by Ehtesham-Zahid** [comment]
I completely agree that high-level summaries are essential to making the review process sustainable. @laurent22 , to address this, I’ve been prototyping automated reports using Semgrep and Socket.dev as part of my GSoC research.

These tools are designed to explicitly flag new permissions, hidden eval() calls, and supply-chain changes in a structured diff, ensuring that critical updates in package.json or source files aren't overlooked during a review. I’ve detailed this 'Security-First Diff' workflow in my proposal and would love to see if this matches your vision for the review infrastructure!

Proposal Link: https://discourse.joplinapp.org/t/gsoc-2026-proposal-draft-idea-6-strengthen-the-security-of-the-plugin-ecosystem-ehtesham-zahid/49224/11
